File: Data Structures & Algorithms/palindrome-partitioning/submission-2.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        dp = [[False for l in range(r+1)] for r in range(len(s))]
        for r in range(len(s)):
            for l in range(r+1):
                if (l == r or 
                    (s[l] == s[r] and r - l <= 2) or 
                    (s[l] == s[r] and dp[r-1][l+1])):
                    dp[r][l] = True
        
        # print(dp)

        res = []
        stack = [(0, [])]
        while stack:
            l, path = stack.pop()
            if l == len(s):
                res.append(path)
            else:
                for r in range(l, len(s)):
                    if dp[r][l]:
                        stack.append((r + 1, path + [s[l:r + 1]]))
        return res

File: Data Structures & Algorithms/course-schedule-ii/submission-2.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        // Indegree-based topological sort
        std::vector<int> nPrereqs(numCourses);
        std::vector<std::vector<int>> satisfies(numCourses);
        for (const auto &item : prerequisites) {
            ++(nPrereqs[item[0]]);
            satisfies[item[1]].push_back(item[0]);
        }

        std::queue<int> q;
        for (int i = 0; i < numCourses; ++i) {
            if (nPrereqs[i] == 0) q.push(i);
        }

        std::vector<int> path; path.reserve(numCourses);
        while (!q.empty()) {
            int curr = q.front(); q.pop(); path.push_back(curr);
            for (int next : satisfies[curr]) {
                --(nPrereqs[next]);
                if (nPrereqs[next] == 0) {
                    q.push(next);
                }
            }
        }

        return (path.size() == numCourses) ? path : vector<int>{};
    }
};
